promql/test.go

```diff
@@ -584,67 +733,15 @@ func (t *test) exec(tc testCommand, engine engineQuerier) error {
 		}
 		queries = append([]atModifierTestCase{{expr: cmd.expr, evalTime: cmd.start}}, queries...)
 		for _, iq := range queries {
-			q, err := engine.NewInstantQuery(t.context, t.storage, nil, iq.expr, iq.evalTime)
-			if err != nil {
-				return err
-			}
-			defer q.Close()
-			res := q.Exec(t.context)
-			if res.Err != nil {
-				if cmd.fail {
-					continue
-				}
-				return fmt.Errorf("error evaluating query %q (line %d): %w", iq.expr, cmd.line, res.Err)
-			}
-			if res.Err == nil && cmd.fail {
-				return fmt.Errorf("expected error evaluating query %q (line %d) but got none", iq.expr, cmd.line)
-			}
-			err = cmd.compareResult(res.Value)
-			if err != nil {
-				return fmt.Errorf("error in %s %s (line %d): %w", cmd, iq.expr, cmd.line, err)
-			}
-
-			// Check query returns same result in range mode,
-			// by checking against the middle step.
-			q, err = engine.NewRangeQuery(t.context, t.storage, nil, iq.expr, iq.evalTime.Add(-time.Minute), iq.evalTime.Add(time.Minute), time.Minute)
-			if err != nil {
+			if err := t.runQuery(iq, cmd, engine); err != nil {
 				return err
 			}
-			rangeRes := q.Exec(t.context)
-			if rangeRes.Err != nil {
-				return fmt.Errorf("error evaluating query %q (line %d) in range mode: %w", iq.expr, cmd.line, rangeRes.Err)
-			}
-			defer q.Close()
-			if cmd.ordered {
-				// Ordering isn't defined for range queries.
-				continue
-			}
-			mat := rangeRes.Value.(Matrix)
-			vec := make(Vector, 0, len(mat))
-			for _, series := range mat {
-				// We expect either Floats or Histograms.
-				for _, point := range series.Floats {
-					if point.T == timeMilliseconds(iq.evalTime) {
-						vec = append(vec, Sample{Metric: series.Metric, T: point.T, F: point.F})
-						break
-					}
-				}
-				for _, point := range series.Histograms {
-					if point.T == timeMilliseconds(iq.evalTime) {
-						vec = append(vec, Sample{Metric: series.Metric, T: point.T, H: point.H})
-						break
-					}
+			if strings.Contains(iq.expr, "_bucket") {
```

Comment:
I'd rather we didn't do this, we should make the copy by hand to be explicit. Unit tests should be very clear, explicit and define expectations.

Reply:
Technically we have 5 choices as far as I see:
- explicit tests (easy, but error prone)
- assert that all tests have both versions
- able to mark tests to include in checks with nhcb
- able to mark tests to exclude in checks with nhcb
- do what we do now and fix all cases (not possible)

Do we have access to all `eval` statements? We could avoid having to mark exceptions by just defining the nhcb version explicitly.

Also technically you'd have to remove `le` label from the nhcb expressions as that's the official conversion to native query.

Reply:
The 2nd one seems a bit complicated. 3rd is slightly more troublesome to implement than the 4th but it is more explicit, I don't really mind 3rd or 4th.

> We could avoid having to mark exceptions by just defining the nhcb version explicitly.
Sorry, what do you mean by this?

> Also technically you'd have to remove le label from the nhcb expressions as that's the official conversion to native query.
Yes it works well even without doing it, but it should be simple enough to just remove it as well.

Reply:
```
# More realistic with rates.
eval instant at 50m histogram_quantile(0.2, rate(testhistogram_bucket[5m]))
	{start="positive"} 0.048
	{start="negative"} -0.2

eval instant at 50m histogram_quantile(0.5, rate(testhistogram_bucket[5m]))
	{start="positive"} 0.15
	{start="negative"} -0.15

eval instant at 50m histogram_quantile(0.8, rate(testhistogram_bucket[5m]))
	{start="positive"} 0.72
	{start="negative"} 0.3
```

The above (originally included) tests were initially failing only for the nhcb version, but the bug is now fixed. Note that even without the fix, the following tests were working for both classic and nhcb:

```
# More realistic with rates.
eval instant at 50m histogram_quantile(0.2, rate(testhistogram_bucket{start="positive"}[5m]))
	{start="positive"} 0.048

eval instant at 50m histogram_quantile(0.2, rate(testhistogram_bucket{start="negative"}[5m]))
	{start="negative"} -0.2

eval instant at 50m histogram_quantile(0.5, rate(testhistogram_bucket{start="positive"}[5m]))
	{start="positive"} 0.15

eval instant at 50m histogram_quantile(0.5, rate(testhistogram_bucket{start="negative"}[5m]))
	{start="negative"} -0.15

eval instant at 50m histogram_quantile(0.8, rate(testhistogram_bucket{start="positive"}[5m]))
	{start="positive"} 0.72

eval instant at 50m histogram_quantile(0.8, rate(testhistogram_bucket{start="negative"}[5m]))
	{start="negative"} 0.3
```

Meaning that when you mix the positive and negative series (that have different bucket bounds) together, something gets corrupted somewhere. This isn't a problem for the other tests as the different series have consistent bucket bounds. This isn't a problem for classic histograms or the exponential native histograms, so the problem is with our new customValues field, which we shallow copied when we added it to the model instead of doing a deep copy. I added the fix commit to the earlier chunks PR because I'd rather get the fix in main earlier. We can revisit these optimisations later on when nhcb implementation is more complete and we have more tests.

Reply:
Moved this part to https://github.com/prometheus/prometheus/pull/13827

Reply:
Thinking more about the test setup, I'm more and more inclined to make everything explicit.

For one, that means that `load`ing a classic histogram needs to have some keyword to activate conversion into NHCB, mostly because there are cases where the test deliberately contains data that cannot be converted to NHCB (see above), _unless_ we decide to tweak the conversion algorithm so that it can deal with those cases (which maybe we should not, see also above). (I think this is an agreement with @zenador's assessment so far.)

However, I also think we should give up on running the classic histogram tests automatically and implicitly with converted NHCB. Here are the reasons I see:
- There are queries that deliberately create a different result, e.g. `rate(histo_bucket)` creates a classic histogram as output, while `rate(histo)` creates a native histogram. There are more subtle cases like the counter reset detection that is deliberately broken in classic histograms with negative observations in the game but works correctly after NHCB conversion. (I don't know if we have a test for that right now, but we could one day or maybe even should.)
- Even for queries that are supposed to create the same output, the conversion isn't trivial. `histogram_quantile(0.5, histo_bucket)` is easy. `histogram_quantile(0.5, sum by (le) (histo_bucket))` is a bit harder. You also have to take into account the also valid `histogram_quantile(0.5, sum(histo_bucket) by (le))` (and there might be more labels in any order and you have to find the `le` label and remove it). Then there are queries involving `foo_count` and `foo_sum`, which needed to be converted to `histogram_count(foo)` and `histogram_sum(foo)`. Etc. In short, I think implementing a correct test converter is a very deep rabbit hole.
- And now add the problem that there are cases where the conversion to NHCB doesn't work by design (see above, where we concluded we need to mark the conversion explicitly anyay).

In summary, I think would need to invest a lot of effort to chase an ultimately impossible goal if we desire to automatically run all classic histogram tests with NHCB, too. In many cases, we need a human anyway to decide if the test should run with NHCB at all, and if yes how the query will look like.

Reply:
> maybe even one where the same buckets are defined with different formats (like now) but they have different values. (I'm not even sure what the behavior should be then.

Yeah, it seems like in that case we could arbitrarily select one and the behaviour would be non-deterministic, so perhaps it's better not to test that and just say we provide no guarantees in such edge cases.

> I would currently tend towards (1).

I agree, it is better to only convert valid histograms. If they are problematic, better to flag it out earlier with an error so the user can look into fixing it, than to automatically correct it upon conversion based on assumptions that may be wrong.

Anyway, now both loading and evaluating with native histograms are opt-in, you need to use the keyword suffix `_with_nhcb` for both of those. So we don't have those suffixes for the `nonmonotonic_bucket`, and for `mixed_bucket` we load it, use the keyword for tests with the same results, and don't use the keyword for the failing test, and we now have a separate test for the different result of the nhcb version, so when this disparity gets looked into in future, the tests can be updated to indicate the change in behaviour.

The conversion in the eval query is very simple, and it works well for the current tests that we have, but for more complicated cases like the one you mentioned, it's best that we don't use the keyword and define a separate test instead. I agree that making the test converter completely correct would be more trouble than it's worth, so we'll just have a simple one for convenience to cover most of the simple cases, and fall back to manually defining it when necessary.